go/merklestore/store.go

```diff
@@ -0,0 +1,346 @@
+		return ret, NewMerkleStoreError("empty %s for version: %d", s.tag, s.supportedVersion)
+	}
+
+	return keybase1.MerkleStoreEntry{
+		Hash:  hash,
+		Entry: keybase1.MerkleStoreEntryString(sub),
+	}, nil
+}
+
+// Get stored kit as a string.  First it makes sure that the merkle root is
+// recent enough.  Using the hash from that, it fetches from in-memory falling
+// back to db falling back to server.
+func (s *MerkleStoreImpl) getKitString(m libkb.MetaContext) (
+	keybase1.MerkleStoreKit, keybase1.MerkleStoreKitHash, error) {
+
+	// Use a file instead if specified.
+	if len(s.kitFilename) > 0 {
+		m.CDebugf("MerkleStore: using kit file: %s", s.kitFilename)
+		return s.readFile(s.kitFilename)
+	}
+
+	mc := m.G().GetMerkleClient()
+	if mc == nil {
+		return "", "", NewMerkleStoreError("no MerkleClient available")
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	root := mc.LastRoot()
+	// The time that the root was fetched is used rather than when the
+	// root was published so that we can continue to operate even if
+	// the root has not been published in a long time.
+	if (root == nil) || s.pastDue(m, root.Fetched(), libkb.MerkleStoreShouldRefresh) {
+		m.CDebugf("MerkleStore: merkle root should refresh")
+
+		// Attempt a refresh if the root is old or nil.
+		err := s.refreshRoot(m)
+		if err != nil {
+			m.CDebugf("MerkleStore: could not refresh merkle root: %s", err)
+		} else {
+			root = mc.LastRoot()
+		}
+	}
+
+	if root == nil {
+		return "", "", NewMerkleStoreError("no merkle root")
+	}
+
+	if s.pastDue(m, root.Fetched(), libkb.MerkleStoreRequireRefresh) {
+		// The root is still too old, even after an attempted refresh.
+		m.CDebugf("MerkleStore: merkle root too old")
+		return "", "", NewMerkleStoreError("merkle root too old: %v %s", seqnoWrap(root.Seqno()), root.Fetched())
+	}
+
+	// This is the hash we are being instructed to use.
+	hash := keybase1.MerkleStoreKitHash(s.getRootHash(*root))
+
+	if hash == "" {
+		return "", "", NewMerkleStoreError("merkle root has empty %s hash: %v", s.tag, seqnoWrap(root.Seqno()))
+	}
+
+	// Use in-memory cache if it matches
+	if fromMem := s.memGet(hash); fromMem != nil {
+		m.CDebugf("MerkleStore: mem cache hit, using hash: %s", hash)
+		return *fromMem, hash, nil
+	}
+
+	// Use db cache if it matches
+	if fromDB := s.dbGet(m, hash); fromDB != nil {
+		m.CDebugf("MerkleStore: db cache hit")
+
+		// Store to memory
+		s.memSet(hash, *fromDB)
+
+		m.CDebugf("MerkleStore: using hash: %s", hash)
+		return *fromDB, hash, nil
+	}
+
+	// Fetch from the server
+	// This validates the hash
+	kitJSON, err := s.fetch(m, hash)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Store to memory
+	s.memSet(hash, kitJSON)
+
+	// db write
+	s.dbSet(m.BackgroundWithLogTags(), hash, kitJSON)
+
+	m.CDebugf("MerkleStore: using hash: %s", hash)
+	return kitJSON, hash, nil
+}
+
+type merkleStoreServerRes struct {
+	Status  libkb.AppStatus         `json:"status"`
+	KitJSON keybase1.MerkleStoreKit `json:"kit_json"`
+}
```

Comment:
I didn't realize the applications were so similar that this could be shared. Nice.